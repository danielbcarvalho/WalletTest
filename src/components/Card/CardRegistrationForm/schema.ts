import * as yup from 'yup';
import { isValidCreditCard, isValidExpirationDate } from '../../../utils/card';

export const cardFormValidationSchema = yup.object({
  number: yup
    .string()
    .required('Número do cartão é obrigatório')
    .test('creditCard', 'Número do cartão inválido', value => {
      return value ? isValidCreditCard(value) : true;
    }),
  name: yup
    .string()
    .required('Nome do titular do cartão é obrigatório')
    .test('fullName', 'Nome incompleto', value => {
      return value ? value.trim().split(' ').length > 1 : true;
    }),
  expiry: yup
    .string()
    .required('Data de vencimento é obrigatória')
    .matches(/^\d{2}\/\d{2}$/, 'Data de vencimento inválida')
    .test('expirationDate', 'Data de vencimento inválida', value => {
      return value ? isValidExpirationDate(value) : true;
    }),
  cvv: yup
    .string()
    .required('Código de segurança é obrigatório')
    .matches(/^\d{3}$/, 'Código de segurança inválido'),
});
