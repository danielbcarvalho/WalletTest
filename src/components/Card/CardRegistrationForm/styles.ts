import styled from 'styled-components/native';

export const BottomInputWrapper = styled.TouchableOpacity`
  flex-direction: row;
  justify-content: space-between;
`;

export const ScrollView = styled.ScrollView.attrs({
  showsVerticalScrollIndicator: false,
  contentContainerStyle: { flexGrow: 1 },
})``;
