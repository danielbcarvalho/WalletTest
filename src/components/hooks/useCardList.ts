import { useState } from 'react';
import { AxiosError } from 'axios';
import { useNavigation } from '@react-navigation/native';
import { useQuery, useQueryClient } from '@tanstack/react-query';

import { Card } from '../../models/CardModels';
import { cardService } from '../../services/card';
import { QueryKeys } from '../../services/QueryKeys';
import { isCardOnTheTop, topCardIndex } from '../../utils/card';
import { CardListScreenProps } from '../../screens/CardListScreen';

const FIVE_MINUTES = 1000 * 60 * 5;

export function useCardList() {
  const queryClient = useQueryClient();
  const { navigate } = useNavigation<CardListScreenProps>();
  const [useThisCard, setUseThisCard] = useState<boolean>(false);

  const {
    error,
    isError,
    data: cards,
    isInitialLoading,
  } = useQuery<Card[]>([QueryKeys.CARD_LIST], () => cardService.list(), {
    staleTime: FIVE_MINUTES,
  });

  function setCardOnTop(selectedCardId: string) {
    if (cards) {
      const cardSelectedIndex = cards.findIndex(
        card => card.id === selectedCardId,
      );

      if (!isCardOnTheTop(cardSelectedIndex, cards)) {
        queryClient.setQueryData(
          [QueryKeys.CARD_LIST],
          [cards[topCardIndex(cards)], cards[cardSelectedIndex]],
        );
      }
    }
  }

  function onUseThisCard() {
    setUseThisCard(true);
    navigate('CardPayment');
  }

  return {
    cards: cards ?? [],
    error: error as AxiosError | undefined,
    isError,
    useThisCard,
    setCardOnTop,
    onUseThisCard,
    isLoading: isInitialLoading,
  };
}
