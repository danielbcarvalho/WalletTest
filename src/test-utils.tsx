// TODO: Verificar essa melhoria nos testes
import React from 'react';
import {
  render,
  RenderAPI,
  RenderOptions,
} from '@testing-library/react-native';
import renderer, { ReactTestRenderer } from 'react-test-renderer';
import { ThemeProvider } from 'styled-components/native';

import { mockDefaultTheme } from './theme/defaultTheme';

interface IAllThemeProviders {
  children: React.ReactElement;
}
const AllTheProviders = ({ children }: IAllThemeProviders) => {
  return <ThemeProvider theme={mockDefaultTheme}>{children}</ThemeProvider>;
};

function customRender(
  ui: React.ReactElement,
  options = {} as RenderOptions,
): RenderAPI {
  return render(ui, { wrapper: AllTheProviders, ...options });
}

export function rendererCreate(
  component: React.ReactElement,
): ReactTestRenderer {
  return renderer.create(<AllTheProviders>{component}</AllTheProviders>);
}

// re-export everything
export * from '@testing-library/react-native';

// override render method
export { customRender as render };
