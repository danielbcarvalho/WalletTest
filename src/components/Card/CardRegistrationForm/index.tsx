import React from 'react';

import { useTranslation } from 'react-i18next';
import { Masks } from 'react-native-mask-input';
import { yupResolver } from '@hookform/resolvers/yup';
import { useForm, Controller } from 'react-hook-form';
import { useFocusEffect } from '@react-navigation/native';

import { Button } from '../../Button';
import InputCustom from '../../InputCustom';

import { BottomInputWrapper } from './styles';
import { cardFormValidationSchema } from './schema';
import { getMaskCVV, getMaskDateMMYY } from '../../../utils/mask-input';

interface Props {
  onCardRegisterForm: (card: FormValues) => void;
}

interface FormValues {
  number: string;
  name: string;
  expiry: string;
  cvv: string;
}

function CardRegistrationForm({ onCardRegisterForm }: Props) {
  const { t } = useTranslation();
  const {
    control,
    handleSubmit,
    formState,
    reset: clearForm,
  } = useForm<FormValues>({
    defaultValues: {
      number: '',
      name: '',
      expiry: '',
      cvv: '',
    },
    resolver: yupResolver(cardFormValidationSchema),
    mode: 'all',
  });

  const onSubmit = (data: FormValues) => {
    const card: FormValues = {
      name: data.name,
      number: data.number,
      cvv: data.cvv,
      expiry: data.expiry,
    };
    onCardRegisterForm(card);
  };

  useFocusEffect(
    React.useCallback(() => {
      const unsubscribe = () => clearForm();

      return () => unsubscribe();
    }, [clearForm]),
  );

  return (
    <>
      <Controller
        control={control}
        render={({ field: { onChange, onBlur, value } }) => {
          return (
            <InputCustom
              accessibilityLabel={t('card number')}
              onBlur={onBlur}
              onChangeText={onChange}
              value={value}
              label={t('card number')}
              placeholder=""
              keyboardType="numeric"
              mask={Masks.CREDIT_CARD}
              name="number"
            />
          );
        }}
        name="number"
      />
      <Controller
        control={control}
        render={({ field: { onChange, onBlur, value } }) => (
          <InputCustom
            accessibilityLabel={t('card holder name')}
            onBlur={onBlur}
            onChangeText={onChange}
            value={value}
            label={t('card holder name')}
            keyboardType="default"
            autoCapitalize="words"
            name="name"
          />
        )}
        name="name"
      />
      <BottomInputWrapper>
        <Controller
          control={control}
          render={({ field: { onChange, onBlur, value } }) => (
            <InputCustom
              accessibilityLabel={t('expiration date')}
              placeholder="00/00"
              onBlur={onBlur}
              onChangeText={onChange}
              value={value}
              label={t('expiration date')}
              keyboardType="numeric"
              width={48}
              mask={getMaskDateMMYY}
              name="expiry"
            />
          )}
          name="expiry"
        />
        <Controller
          control={control}
          render={({ field: { onChange, onBlur, value } }) => (
            <InputCustom
              accessibilityLabel={t('cvv')}
              placeholder="***"
              onBlur={onBlur}
              onChangeText={onChange}
              value={value}
              label={t('cvv')}
              keyboardType="numeric"
              width={48}
              mask={getMaskCVV}
              name="cvv"
            />
          )}
          name="cvv"
        />
      </BottomInputWrapper>
      <Button
        title={t('advance')}
        accessibilityLabel={t('advance')}
        onPress={handleSubmit(onSubmit)}
        disabled={!formState.isValid}
      />
    </>
  );
}

export default CardRegistrationForm;
