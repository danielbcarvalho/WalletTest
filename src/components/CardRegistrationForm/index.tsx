import React from 'react';

import { Masks } from 'react-native-mask-input';
import { yupResolver } from '@hookform/resolvers/yup';
import { useForm, Controller } from 'react-hook-form';

import { Button } from '../Button';
import InputCustom from '../InputCustom';
import { CardProps } from '../../screens/CardRegistrationScreen';

import { BottomInputWrapper } from './styles';
import { cardFormValidationSchema } from './schema';
import { getMaskCVV, getMaskDateMMYY } from '../../utils/mask-input';
import { useTranslation } from 'react-i18next';

interface Props {
  onCardRegisterForm: (card: CardProps) => void;
}

interface FormValues {
  card: string;
  name: string;
  date: string;
  cvv: string;
}

function CardRegistrationForm({ onCardRegisterForm }: Props) {
  const { t } = useTranslation();
  const { control, handleSubmit, formState } = useForm<FormValues>({
    defaultValues: {
      card: '',
      name: '',
      date: '',
      cvv: '',
    },
    resolver: yupResolver(cardFormValidationSchema),
  });

  const onSubmit = (data: FormValues) => {
    const card: CardProps = {
      name: data.name,
      card: data.card,
      cvv: data.cvv,
      date: data.date,
    };
    onCardRegisterForm(card);
  };

  return (
    <>
      <Controller
        control={control}
        render={({ field: { onChange, onBlur, value } }) => (
          <InputCustom
            accessibilityLabel={t('card number')}
            onBlur={onBlur}
            onChangeText={onChange}
            value={value}
            label={t('card number')}
            placeholder=""
            keyboardType="numeric"
            mask={Masks.CREDIT_CARD}
          />
        )}
        name="card"
      />
      <Controller
        control={control}
        render={({ field: { onChange, onBlur, value } }) => (
          <InputCustom
            accessibilityLabel={t('card holder name')}
            onBlur={onBlur}
            onChangeText={onChange}
            value={value}
            label={t('card holder name')}
            keyboardType="default"
            autoCapitalize="words"
          />
        )}
        name="name"
      />
      <BottomInputWrapper>
        <Controller
          control={control}
          render={({ field: { onChange, onBlur, value } }) => (
            <InputCustom
              accessibilityLabel={t('expiration date')}
              placeholder="00/00"
              onBlur={onBlur}
              onChangeText={onChange}
              value={value}
              label={t('expiration date')}
              keyboardType="numeric"
              width={48}
              mask={getMaskDateMMYY}
            />
          )}
          name="date"
        />
        <Controller
          control={control}
          render={({ field: { onChange, onBlur, value } }) => (
            <InputCustom
              accessibilityLabel={t('cvv')}
              placeholder="***"
              onBlur={onBlur}
              onChangeText={onChange}
              value={value}
              label={t('cvv')}
              keyboardType="numeric"
              width={48}
              mask={getMaskCVV}
            />
          )}
          name="cvv"
        />
      </BottomInputWrapper>
      <Button
        title={t('advance')}
        accessibilityLabel={t('advance')}
        onPress={handleSubmit(onSubmit)}
        disabled={!formState.isValid}
      />
    </>
  );
}

export default CardRegistrationForm;
