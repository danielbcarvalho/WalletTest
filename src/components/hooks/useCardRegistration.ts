import uuid from 'react-native-uuid';
import { useTranslation } from 'react-i18next';
import { useNavigation } from '@react-navigation/native';
import { useMutation, useQueryClient } from '@tanstack/react-query';
import { NativeStackNavigationProp } from '@react-navigation/native-stack';

import { ShowAlert } from '../ShowAlert';
import { ParamsList } from '../../Routes';
import { Card } from '../../models/CardModels';
import { cardService } from '../../services/card';
import { QueryKeys } from '../../services/QueryKeys';

import { useCardList } from './useCardList';

type CardRegistrationScreenProps = NativeStackNavigationProp<
  ParamsList,
  'CardRegistration'
>;

interface CardRegistrationFormData {
  cvv: string;
  name: string;
  number: string;
  expiry: string;
}

export function useCardRegistration() {
  const { t } = useTranslation();
  const navigation = useNavigation<CardRegistrationScreenProps>();
  const queryClient = useQueryClient();

  const { cards } = useCardList();

  const registerCardMutation = useMutation(cardService.register, {
    onSuccess: data => {
      queryClient.invalidateQueries([QueryKeys.CARD_LIST]);
      navigation.push('Home');
      navigation.navigate('CardRegistrationSuccess', { card: data });
    },
    onError: (error: any) => {
      ShowAlert({
        title: 'Error',
        message: error.message,
        onPress: () => navigation.navigate('Home'),
      });
    },
  });

  const handleCardRegisterForm = async (data: CardRegistrationFormData) => {
    try {
      if (cards.length >= 2) {
        ShowAlert({
          title: t('Ops, something went wrong'),
          message: t('Soon you will be able to register more cards'),
          onPress: () => navigation.navigate('Home'),
        });
      } else {
        const cardData: Card = {
          id: uuid.v4() as string,
          kind: 'green',
          ...data,
        };

        await registerCardMutation.mutateAsync(cardData);
      }
    } catch (error: any) {
      ShowAlert({
        title: t('Ops, something went wrong'),
        message: t('Try again later'),
        onPress: () => navigation.navigate('Home'),
      });
    }
  };

  return handleCardRegisterForm;
}
