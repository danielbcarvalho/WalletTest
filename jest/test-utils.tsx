import React, { ReactElement, FC } from 'react';
import { render } from '@testing-library/react-native';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { ThemeProvider } from '../src/theme';

type Options = Parameters<typeof render>[1];

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      cacheTime: Infinity,
    },
  },
});
const AllTheProviders: FC<{ children: React.ReactNode }> = ({ children }) => (
  <ThemeProvider>
    <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
  </ThemeProvider>
);

const customRender = (ui: ReactElement, options?: Options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react-native';
export { customRender as render };
