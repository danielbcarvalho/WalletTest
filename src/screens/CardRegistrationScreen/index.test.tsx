import React from 'react';

import { act } from 'react-test-renderer';
import { useNavigation } from '@react-navigation/native';
import { fireEvent, waitFor } from '@testing-library/react-native';

import { render } from '../../../jest/test-utils';
import CardRegistrationForm from '../../components/Card/CardRegistrationForm';
import { useCardRegistration } from '../../components/hooks/useCardRegistration';

import CardRegistration from './';

jest.mock('../../components/hooks/useCardRegistration');
jest.mock('@react-navigation/native', () => ({
  useNavigation: jest.fn(),
}));
describe('CardRegistrationScreen', () => {
  let navigateMock: jest.Mock;

  beforeEach(() => {
    navigateMock = jest.fn();
    useNavigation.mockReturnValue({ navigate: navigateMock });
  });

  test('call useCardRegistration hook and trigger onCardRegisterForm on form submission', async () => {
    const mockOnCardRegisterForm = jest.fn();
    useCardRegistration.mockReturnValue(mockOnCardRegisterForm);

    const { getByLabelText, getByText } = render(
      <CardRegistrationForm onCardRegisterForm={mockOnCardRegisterForm} />,
    );

    await act(async () => {
      fireEvent.changeText(
        getByLabelText('card number'),
        '5284 7474 3246 1715',
      );
      fireEvent.changeText(getByLabelText('card holder name'), 'John Doe');
      fireEvent.changeText(getByLabelText('expiration date'), '12/23');
      fireEvent.changeText(getByLabelText('cvv'), '123');
    });

    await act(async () => {
      fireEvent.press(getByText('advance'));
    });
    await waitFor(() =>
      expect(mockOnCardRegisterForm).toHaveBeenCalledTimes(1),
    );
  });

  test('matches snapshot', () => {
    const { toJSON } = render(<CardRegistration />);
    expect(toJSON()).toMatchSnapshot();
  });
});
