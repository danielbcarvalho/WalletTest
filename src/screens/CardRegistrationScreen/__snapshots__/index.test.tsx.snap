// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardRegistrationScreen matches snapshot 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "backgroundColor": "#142995",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
      {
        "paddingBottom": 0,
      },
    ]
  }
>
  <RCTScrollView
    contentContainerStyle={
      {
        "flexGrow": 1,
      }
    }
    keyboardShouldPersistTaps="handled"
    style={
      {
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      }
    }
  >
    <View>
      <RNCSafeAreaView
        edges={
          {
            "bottom": "additive",
            "left": "additive",
            "right": "additive",
            "top": "additive",
          }
        }
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#142995",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
            "paddingBottom": 10,
            "paddingLeft": 30,
            "paddingRight": 30,
            "paddingTop": 10,
          }
        }
      >
        <View
          style={
            {
              "backgroundColor": "#EEEEEE",
              "borderBottomLeftRadius": 50,
              "borderBottomRightRadius": 50,
              "height": 230,
              "opacity": 0.2,
              "position": "absolute",
              "right": 120,
              "top": -70,
              "transform": [
                {
                  "rotate": "-39.93deg",
                },
              ],
              "width": 370,
            }
          }
          top={true}
        />
        <View
          style={
            {
              "backgroundColor": "#EEEEEE",
              "borderBottomLeftRadius": 50,
              "borderBottomRightRadius": 50,
              "bottom": -70,
              "height": 230,
              "left": 120,
              "opacity": 0.2,
              "position": "absolute",
              "transform": [
                {
                  "rotate": "139deg",
                },
              ],
              "width": 370,
            }
          }
        />
        <Text
          style={
            {
              "color": "#FFF",
              "fontFamily": "PTSans-Caption",
              "fontSize": 28,
              "lineHeight": 52,
            }
          }
        >
          Wallet Test
        </Text>
        <View
          style={
            {
              "marginBottom": 16,
              "marginLeft": 10,
              "marginRight": 10,
              "marginTop": 16,
              "width": "100%",
            }
          }
        >
          <Text
            style={
              {
                "color": "#FFFFFF",
                "fontFamily": "PTSans-Regular",
                "fontSize": 14,
                "marginBottom": 6,
              }
            }
          >
            card number
          </Text>
          <TextInput
            accessibilityLabel="card number"
            autoComplete="off"
            autoCorrect={false}
            isDirty={false}
            keyboardType="numeric"
            onBlur={[Function]}
            onChangeText={[Function]}
            onSelectionChange={[Function]}
            placeholder=""
            style={
              {
                "backgroundColor": "#EEEEEE",
                "borderBottomColor": "transparent",
                "borderBottomLeftRadius": 6,
                "borderBottomRightRadius": 6,
                "borderColor": "transparent",
                "borderLeftColor": "transparent",
                "borderRightColor": "transparent",
                "borderStyle": "solid",
                "borderTopColor": "transparent",
                "borderTopLeftRadius": 6,
                "borderTopRightRadius": 6,
                "borderWidth": 2,
                "color": "#111",
                "fontFamily": "PTSans-Regular",
                "fontSize": 16,
                "height": 45,
                "marginTop": 4,
                "overflow": "hidden",
                "paddingBottom": 0,
                "paddingLeft": 16,
                "paddingRight": 16,
                "paddingTop": 0,
                "width": "100%",
              }
            }
            value=""
          />
        </View>
        <View
          style={
            {
              "marginBottom": 16,
              "marginLeft": 10,
              "marginRight": 10,
              "marginTop": 16,
              "width": "100%",
            }
          }
        >
          <Text
            style={
              {
                "color": "#FFFFFF",
                "fontFamily": "PTSans-Regular",
                "fontSize": 14,
                "marginBottom": 6,
              }
            }
          >
            card holder name
          </Text>
          <TextInput
            accessibilityLabel="card holder name"
            autoCapitalize="words"
            autoComplete="off"
            autoCorrect={false}
            isDirty={false}
            keyboardType="default"
            onBlur={[Function]}
            onChangeText={[Function]}
            onSelectionChange={[Function]}
            style={
              {
                "backgroundColor": "#EEEEEE",
                "borderBottomColor": "transparent",
                "borderBottomLeftRadius": 6,
                "borderBottomRightRadius": 6,
                "borderColor": "transparent",
                "borderLeftColor": "transparent",
                "borderRightColor": "transparent",
                "borderStyle": "solid",
                "borderTopColor": "transparent",
                "borderTopLeftRadius": 6,
                "borderTopRightRadius": 6,
                "borderWidth": 2,
                "color": "#111",
                "fontFamily": "PTSans-Regular",
                "fontSize": 16,
                "height": 45,
                "marginTop": 4,
                "overflow": "hidden",
                "paddingBottom": 0,
                "paddingLeft": 16,
                "paddingRight": 16,
                "paddingTop": 0,
                "width": "100%",
              }
            }
            value=""
          />
        </View>
        <View
          style={
            {
              "flexDirection": "row",
              "justifyContent": "space-between",
            }
          }
        >
          <View
            style={
              {
                "marginBottom": 16,
                "marginLeft": 10,
                "marginRight": 10,
                "marginTop": 16,
                "width": "48%",
              }
            }
            width={48}
          >
            <Text
              style={
                {
                  "color": "#FFFFFF",
                  "fontFamily": "PTSans-Regular",
                  "fontSize": 14,
                  "marginBottom": 6,
                }
              }
            >
              expiration date
            </Text>
            <TextInput
              accessibilityLabel="expiration date"
              autoComplete="off"
              autoCorrect={false}
              isDirty={false}
              keyboardType="numeric"
              onBlur={[Function]}
              onChangeText={[Function]}
              onSelectionChange={[Function]}
              placeholder="00/00"
              style={
                {
                  "backgroundColor": "#EEEEEE",
                  "borderBottomColor": "transparent",
                  "borderBottomLeftRadius": 6,
                  "borderBottomRightRadius": 6,
                  "borderColor": "transparent",
                  "borderLeftColor": "transparent",
                  "borderRightColor": "transparent",
                  "borderStyle": "solid",
                  "borderTopColor": "transparent",
                  "borderTopLeftRadius": 6,
                  "borderTopRightRadius": 6,
                  "borderWidth": 2,
                  "color": "#111",
                  "fontFamily": "PTSans-Regular",
                  "fontSize": 16,
                  "height": 45,
                  "marginTop": 4,
                  "overflow": "hidden",
                  "paddingBottom": 0,
                  "paddingLeft": 16,
                  "paddingRight": 16,
                  "paddingTop": 0,
                  "width": "100%",
                }
              }
              value=""
            />
          </View>
          <View
            style={
              {
                "marginBottom": 16,
                "marginLeft": 10,
                "marginRight": 10,
                "marginTop": 16,
                "width": "48%",
              }
            }
            width={48}
          >
            <Text
              style={
                {
                  "color": "#FFFFFF",
                  "fontFamily": "PTSans-Regular",
                  "fontSize": 14,
                  "marginBottom": 6,
                }
              }
            >
              cvv
            </Text>
            <TextInput
              accessibilityLabel="cvv"
              autoComplete="off"
              autoCorrect={false}
              isDirty={false}
              keyboardType="numeric"
              onBlur={[Function]}
              onChangeText={[Function]}
              onSelectionChange={[Function]}
              placeholder="***"
              style={
                {
                  "backgroundColor": "#EEEEEE",
                  "borderBottomColor": "transparent",
                  "borderBottomLeftRadius": 6,
                  "borderBottomRightRadius": 6,
                  "borderColor": "transparent",
                  "borderLeftColor": "transparent",
                  "borderRightColor": "transparent",
                  "borderStyle": "solid",
                  "borderTopColor": "transparent",
                  "borderTopLeftRadius": 6,
                  "borderTopRightRadius": 6,
                  "borderWidth": 2,
                  "color": "#111",
                  "fontFamily": "PTSans-Regular",
                  "fontSize": 16,
                  "height": 45,
                  "marginTop": 4,
                  "overflow": "hidden",
                  "paddingBottom": 0,
                  "paddingLeft": 16,
                  "paddingRight": 16,
                  "paddingTop": 0,
                  "width": "100%",
                }
              }
              value=""
            />
          </View>
        </View>
        <View
          accessibilityLabel="advance"
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": true,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "alignSelf": "stretch",
              "backgroundColor": "#EEEEEE",
              "borderBottomLeftRadius": 12,
              "borderBottomRightRadius": 12,
              "borderTopLeftRadius": 12,
              "borderTopRightRadius": 12,
              "justifyContent": "center",
              "marginBottom": 10,
              "marginLeft": 0,
              "marginRight": 0,
              "marginTop": 10,
              "maxHeight": 56,
              "minHeight": 56,
              "opacity": 1,
              "paddingBottom": 10,
              "paddingLeft": 20,
              "paddingRight": 20,
              "paddingTop": 10,
            }
          }
          testID="multipurpose-button"
        >
          <Text
            style={
              {
                "color": "#BBBBBB",
                "fontFamily": "PTSans-Regular",
                "fontSize": 18,
              }
            }
            type="disabled"
          >
            advance
          </Text>
        </View>
      </RNCSafeAreaView>
    </View>
  </RCTScrollView>
</View>
`;
